core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -683,6 +705,92 @@ class half_t {
     return __half2float(val) >= __half2float(rhs.val);
 #endif
   }
+
+  KOKKOS_FUNCTION
+  friend bool operator==(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) ==
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) ==
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator!=(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) !=
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) !=
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator<(const volatile half_t& lhs,
+                        const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) <
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) <
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator>(const volatile half_t& lhs,
+                        const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) >
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) >
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator<=(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) <=
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) <=
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  KOKKOS_FUNCTION
+  friend bool operator>=(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<impl_type&>(lhs.val) >=
+                             const_cast<impl_type&>(rhs.val));
+#else
+    return __half2float(const_cast<impl_type&>(lhs.val)) >=
+           __half2float(const_cast<impl_type&>(rhs.val));
+#endif
+  }
+
+  // Insertion and extraction operators
+  friend std::ostream& operator<<(std::ostream& os, const half_t& x) {
+    const std::string out = std::to_string(static_cast<double>(x));
```

Comment:
```
var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Half.hpp:783:23: error: implicit instantiation of undefined template 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>' [clang-diagnostic-error]
    const std::string out = std::to_string(static_cast<double>(x));
                      ^
/usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stringfwd.h:71:11: note: template is declared here
    class basic_string;
          ^
/var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Half.hpp:783:34: error: no member named 'to_string' in namespace 'std' [clang-diagnostic-error]
    const std::string out = std::to_string(static_cast<double>(x));
                                 ^
/var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Half.hpp:789:17: error: implicit instantiation of undefined template 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>' [clang-diagnostic-error]
    std::string in;
                ^
```
Looks like you are missing to include `<string>`.

Reply:
Thanks! Any idea what's going on in the appveyor build?

Reply:
That's fixed in https://github.com/kokkos/kokkos/pull/4102.